Mathlib/Analysis/SpecialFunctions/BinEntropy.lean

```diff
@@ -0,0 +1,508 @@
+/-
+Copyright (c) 2023 Adomas Baliuka. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adomas Baliuka
+-/
+import Mathlib.Analysis.SpecialFunctions.Log.NegMulLog
+import Mathlib.Analysis.Convex.SpecificFunctions.Basic
+
+/-!
+# Properties of Shannon q-ary entropy and binary entropy functions
+
+The [binary entropy function](https://en.wikipedia.org/wiki/Binary_entropy_function)
+`binaryEntropy p := - p * log p - (1 - p) * log (1 - p)`
+is the Shannon entropy of a Bernoulli random variable with success probability `p`.
+
+More generally, the q-ary entropy function is the Shannon entropy of the random variable
+with possible outcomes `{1, ..., q}`, where outcome `1` has probability `(1 - p)`
+and all other outcomes are equally likely.
+
+`qaryEntropy (q : ℕ) (p : ℝ) := p * log (q - 1) - p * log p - (1 - p) * log (1 - p)`
+
+This file assumes that entropy is measured in Nats, hence the use of natural logarithms.
+Most lemmas are also valid using a logarithm in a different base.
+
+## Tags
+
+entropy, Shannon, binary
+-/
+
+open Real
+
+/-- Shannon q-ary Entropy function (measured in Nats, i.e., using natural logs).
+
+It's the Shannon entropy of a random variable with possible outcomes {1, ..., q}
+where outcome `1` has probability `(1 - p)` and all other outcomes are equally likely.
+
+Usual domain of definition is p ∈ [0,1], i.e., input is a probability.
+
+This is a generalization of the binary entropy function `binaryEntropy`. -/
+noncomputable def qaryEntropy (q : ℕ) (p : ℝ) : ℝ :=
+  p * log (q - 1) - p * log p - (1 - p) * log (1 - p)
+
+/-- The [binary entropy function](https://en.wikipedia.org/wiki/Binary_entropy_function)
+`binaryEntropy p := - p * log p - (1-p) * log (1 - p)`
+is the Shannon entropy of a Bernoulli random variable with success probability `p`. -/
+noncomputable def binaryEntropy := qaryEntropy 2
+
+lemma binaryEntropy_eq : binaryEntropy = (fun p => -p * log p - (1 - p) * log (1 - p)) := by
+  have : (2 : ℝ) - 1 = 1 := by norm_num
+  ext
+  simp [binaryEntropy, qaryEntropy, this]
+
+lemma binaryEntropy_eq' {p : ℝ} : binaryEntropy p = -p * log p - (1 - p) * log (1 - p) := by
```

Comment:
```suggestion
lemma binaryEntropy_apply (p : ℝ) : binaryEntropy p = -p * log p - (1 - p) * log (1 - p) := by
```

Reply:
What's the logic behind calling this "apply"?